cooked-validators/src/Cooked/MockChain/Monad/Direct.hs

```diff
@@ -240,58 +251,58 @@ utxoIndex0 = utxoIndex0From def
-  -- -- future
-  -- firstSigner NE.:| _ <- askSigners
-  -- let balancingWallet = firstSigner
-  -- let balancingWalletPkh = walletPKHash firstSigner
-  -- let collateralWallet = firstSigner
-  -- skel <-
-  --   if balance . txSkelOpts $ skelUnbal
-  --     then
-  --       setFeeAndBalance
-  --         balancingWalletPkh
-  --         -- We need to add the balancing wallet to the signers before
-  --         -- calculating the fees. The fees depend on the number of signers: they
-  --         -- make the transaction heavier.
-  --         ( skelUnbal
-  --             & txSkelRequiredSignersL
-  --             %~ (<> Set.singleton balancingWalletPkh)
-  --         )
-  --     else return skelUnbal
-  -- collateralInputs <- calcCollateral collateralWallet
-  -- params <- askParams
-  -- managedData <- gets mcstDatums
-  -- -- See the comment at the arguments of 'runTransactionValidation' on why we
-  -- -- have to use 'generateTxBodyContentWithoutInputDatums' here. The
-  -- -- documentation of the 'generateTxBodyContent*' functions might also be
-  -- -- informative.
-  -- case generateTxBodyContent withoutDatums {gtpCollateralIns = collateralInputs} params managedData skel of
-  --   Left err -> throwError $ MCEGenerationError err
-  --   Right txBodyContent -> do
-  --     slot <- currentSlot
-  --     index <- gets mcstIndex
-  --     someCardanoTx <-
-  --       runTransactionValidation
-  --         slot
-  --         params
-  --         index
-  --         [balancingWallet, collateralWallet]
-  --         txBodyContent
-  --         (txSkelInputData skel)
-  --         (txSkelOutputData skel)
-  --         (unsafeModTx $ txSkelOpts skel)
-  --     when (autoSlotIncrease $ txSkelOpts skel) $
-  --       modify' (\st -> st {mcstCurrentSlot = mcstCurrentSlot st + 1})
-  --     return (Pl.CardanoApiTx someCardanoTx)
-
-  -- txOutByRef outref = gets ((undefined . Pl.getTxOut) <$> Map.lookup outref . Pl.getIndex . mcstIndex)
+  validateTxSkel skelUnbal = do
+    -- TODO We use the first signer as a default wallet for fees and
+    -- collaterals. These should become parameters of validateTxSkel in the
+    -- future
+    firstSigner NE.:| _ <- askSigners
+    let balancingWallet = firstSigner
+    let balancingWalletPkh = walletPKHash firstSigner
+    let collateralWallet = firstSigner
+    skel <-
+      if balance . txSkelOpts $ skelUnbal
+        then
+          setFeeAndBalance
+            balancingWalletPkh
+            -- We need to add the balancing wallet to the signers before
+            -- calculating the fees. The fees depend on the number of signers: they
+            -- make the transaction heavier.
+            ( skelUnbal
+                & txSkelRequiredSignersL
+                %~ (<> Set.singleton balancingWalletPkh)
+            )
+        else return skelUnbal
+    collateralInputs <- calcCollateral collateralWallet -- TODO: Why is it OK to balance first and then add collateral?
```

Comment:
I guess it's because the collateral is never used in the normal execution flow, and collateral is mutually exclusive with the rest of outputs.

At least, that's my intuition that I'm sort of happy with.